lib/spack/spack/solver/asp.py

```diff
@@ -3839,26 +3839,22 @@ def splice_at_hash(
         self._splices.setdefault(parent_node, []).append(splice)
 
     def _resolve_automatic_splices(self):
-        """After all of the specs have been concretized, apply all immediate splices in size order.
+        """After all of the specs have been concretized, apply all immediate splices.
+
+        Use reverse topological order to ensure that all dependencies are resolved
+        before their parents, allowing for maximal sharing and minimal copying.
 
-        This ensures that all dependencies are resolved before their parents, allowing for maximal
-        sharing and minimal copying.
         """
         fixed_specs = {}
-        path_counts = {}
-
-        # Count and cache number of dependency paths in a spec
-        # This proxies for node count and is easier to compute
-        def path_count(spec: spack.spec.Spec):
-            assert spec.concrete
-            count = path_counts.get(id(spec), None)
-            if count is None:
-                count = 1 + sum(path_count(dep) for dep in spec.dependencies())
-                path_counts[id(spec)] = count
-            return count
-
-        # iterate over specs sorted by their dep count
-        for node, spec in sorted(self._specs.items(), key=lambda x: path_count(x[1])):
+        topo_order = [
+            id(spec)
+            for spec in traverse.traverse_nodes(
+                self.specs.values(), order="topo", key=traverse.by_dag_hash
+            )
+        ]
```

Comment:
I think we can use a dict + `enumerate` here, and avoid the call to `index` below. Didn't try the code, but the idea is:
```suggestion
        topo_order = {
            id(spec): index
            for index, spec in enumerate(traverse.traverse_nodes(
                self.specs.values(), order="topo", key=traverse.by_dag_hash
            ))
        }
```

Reply:
nice!